<a href="https://colab.research.google.com/github/DinWaikabu/Pandas/blob/master/mengatasi_missing_value.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np

In [0]:
df = pd.read_csv('http://bit.ly/uforeports')

In [3]:
df.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

### identifikasi missing value

In [4]:
df.isnull().head()

City  Colors Reported  Shape Reported  State   Time
0  False             True           False  False  False
1  False             True           False  False  False
2  False             True           False  False  False
3  False             True           False  False  False
4  False             True           False  False  False

In [5]:
df.isnull().sum() #menhgitung jumlah data missing

City                  25
Colors Reported    15359
Shape Reported      2644
State                  0
Time                   0
dtype: int64

In [6]:
df[df.isnull()].head() #filter dan menampikkan semua data yang memuat missing value

City Colors Reported Shape Reported State Time
0  NaN             NaN            NaN   NaN  NaN
1  NaN             NaN            NaN   NaN  NaN
2  NaN             NaN            NaN   NaN  NaN
3  NaN             NaN            NaN   NaN  NaN
4  NaN             NaN            NaN   NaN  NaN

###Menghapus Missing Value .dropna()

In [9]:
df.shape

(18241, 5)

sebelum dihapus missing value kita memiliki data sebanyak 18241 baris. kemudian baris yang memiliki missing value dihapus dengan asumsi kita memiliki data yang besar dan data yang telah kita hapus dapat kita jumpai di sisa data lainnya

In [0]:
df.dropna(inplace=True) #menhaspus data yang missing

In [11]:
df.isnull().sum() #cek apakah masih terdapat data yang missing

City               0
Colors Reported    0
Shape Reported     0
State              0
Time               0
dtype: int64

In [12]:
df.head()

City Colors Reported Shape Reported State             Time
12      Belton             RED         SPHERE    SC  6/30/1939 20:00
19  Bering Sea             RED          OTHER    AK  4/30/1943 23:00
36  Portsmouth             RED      FORMATION    VA   7/10/1945 1:30
44   Blairsden           GREEN         SPHERE    CA  6/30/1946 19:00
82    San Jose            BLUE        CHEVRON    CA  7/15/1947 21:00

stelah dihapus, maka tidak ada lagi missing value di dalam dataframe yang kita miliki. 

### Interpolasi .fillna()

In [0]:
data = pd.DataFrame([[23, np.nan, 43],
                    [np.nan, 21, 29],
                    [53, 50, 76],
                    [25, 89, np.nan]], columns=['a', 'b', 'c'])

In [14]:
data

a     b     c
0  23.0   NaN  43.0
1   NaN  21.0  29.0
2  53.0  50.0  76.0
3  25.0  89.0   NaN

In [0]:
data.fillna(method='ffill', axis=0, inplace=True) #interpolasi baris

In [19]:
data

a     b     c
0  23.0   NaN  43.0
1  23.0  21.0  29.0
2  53.0  50.0  76.0
3  25.0  89.0  76.0

In [0]:
data.fillna(method='ffill', axis=1, inplace=True) #interpolasi kolom

In [21]:
data

a     b     c
0  23.0  23.0  43.0
1  23.0  21.0  29.0
2  53.0  50.0  76.0
3  25.0  89.0  76.0